saleor/graphql/product/mutations/products.py

```diff
@@ -260,6 +303,20 @@ def clean_input(cls, info, instance, input, errors):
         clean_seo_fields(cleaned_input)
         return cleaned_input
 
+    @classmethod
+    def save(cls, info, instance, cleaned_input):
+        default_variant_data = cleaned_input.get('default_variant', None)
+        instance.save()
+        if default_variant_data:
+            default_variant_instance = models.ProductVariant()
+            default_variant_instance = ProductVariantCreate.construct_instance(
+                default_variant_instance, default_variant_data)
+            default_variant_instance.product = instance
+            default_variant_instance.save()
```

Comment:
We should also trigger somewhere `full_clean` on the newly created variant to validate the model. In the regular ModelMutation flow it's done in the `mutate` function through `clean_instance`. I'm not sure if we will be able to fit this call anywhere in this flow. Since this whole mutation got a bit complex now, maybe it would be much easier to stop using `ModelMutation` here, use `BaseMutation` instead and writing completely custom `mutate` to handle all that logic? We do that for some other complex mutations. What do you think?

Reply:
I've made an attempt to implement all necessary functionality using `ModelMutation`. Frankly I don't really like how `clean_input` of `ProductVariantCreate` is called for `default_variant`. The issue was that `ProductVariantCreate.clean_input` requires `product` to be already created and accessible from variant instance or fetched by its id. On the other hand as I said before implementation of validation logic for `default_variant` separately from `ProductVariantCreate` leads to code duplication. I don't know what's better in this case. How do You think? Or maybe You can advice on another solution